### ISSUE TITLE
Automatic "failover-check" for single checks

### ISSUE THREAD

**Issue by widhalmt** [Icinga/icinga2#6986]
In some setups a check might fail due to network issues, permissions on the client or for other reasons. There is a chance that it will only fail on one node of a HA-pair of nodes.

Since the other node get's informed of failing checks I think it might be rather easy to recheck every not-OK service on the other node. This way we could easily determine if a service is really failing or if it's just some configuration issue on the Icinga node, firewall etc.

I could totally live with the checking becoming "Unkonwn" with a hardcoded or globally configured errormessage like: "Unknown: Check failed on node [nodename] but was ok on node [nodename]"

**Comment by widhalmt** [comment]
ref/NC/603067

**Comment by widhalmt** [comment]
ref/NC/603131

**Comment by Al2Klimov** [comment]
Good catch. I'd prefer the simpler solution of just saying that the check was OK instead of unknown.

However we should discuss this with the other core devs to make sure that it doesn't hurt to hide such check failures. After all your monitoring system shall report if something is wrong.

**Comment by Al2Klimov** [comment]
Just got another idea: Rotate the responsible checker based on the current time.

**Comment by widhalmt** [comment]
Hm. Rotating seems like a good idea but I like the other one better. I think you get more detailed results on first thought.

**Comment by slalomsk8er** [comment]
I would prefer: "WARNING: Check failed on node [nodename] but was ok on node [nodename]" 
Isn't this what warnings are for?

**Comment by widhalmt** [comment]
Nope, @slalomsk8er : You have to differntiate between something being wrong with the service you are monitoring and something being wrong with your monitoring itself.

Critical and Warning indicate, that some problem in the service you are monitoring are imminent.

Unknown means there is a problem within your monitoring system and the status of the monitored object can not be determined with certainty.

**Comment by slalomsk8er** [comment]
@widhalmt: Good point, a differentiation between service, check and Icinga problem would be great. 
In practice I resigned at dozens of unknowns per day so unknowns will not work for us *. Maybe we would catch it in Icingaweb2 because it is not going away like the others but I would definitely miss it in the mails. 

How about a virtual/dynamic/automatic service on the problematic Icinga host itself, called something like icinga_ha_error_service_name?
This way there is a separation between the service and the artificial status that would be injected over the real status. So @Al2Klimov could have his OK because the checked service works. Also the error is for the Icinga admin and not for the service admin - I don't like to pester my customers with my errors.

How to avoid false positives in less then 100% stable environments? Same retry value as the service by default and/or configurable retry value besides the configurable error message?


*Most are SNMP and FTP checks and nobody can tell me if it's the network, the virtual network or the servers - we see the "problem" in 3 different monitoring systems so I think I can rule out my Icinga2 installation.

**Comment by dnsmichi** [comment]
The underlaying logic with self load balancing of [authoritative objects](https://icinga.com/docs/icinga2/latest/doc/19-technical-concepts/#high-availability) on each HA node cannot be changed that easily. You cannot change the object authority on one side, without letting the other node know about and also persisting this over restarts.

If you probably would implement such, how would you prevent flapping between states, e.g. the left node always returns critical, so it sends 1) authority update to master B 2) prepares a modified check result output prior to the moved check 3) awaits check execution and result ... I could imagine that this leads into many broadcast "storms" with more cluster messages in size, cpu processing and also a certain delay of check results dumped into backends.

I would go the route with 2 single checks, and a cluster/business process check on top of it, which actually notifies on XOR/etc. condition.

Cheers,
Michael

**Comment by Al2Klimov** [comment]
IMAO "flapping between states" is exactly what you want in this case as it shows you a defect in your monitoring.

**Comment by Al2Klimov** [comment]
@widhalmt What about pinning two checks, one to each node via #8159 and a third dummy check which compares the results?

**Comment by Al2Klimov** [comment]
_PING_ @widhalmt
